File: alibabacloud-gateway-pop/php/tests/UnitTest.php
<?php

namespace Darabonba\GatewayPop\Tests;

use Darabonba\GatewayPop\Client;
use PHPUnit\Framework\TestCase;

/**
 * @internal
 * @coversNothing
 */
final class UnitTest extends TestCase
{

    public function testGetRegion()
    {
        $client = new Client();
        $this->assertEquals("center", $client->getRegion(null, null, null));
        $this->assertEquals("cn-hangzhou", $client->getRegion(null, null, "cn-hangzhou"));
        $this->assertEquals("center", $client->getRegion("", "", null));
        $this->assertEquals("center", $client->getRegion("test", "test", null));
        $this->assertEquals("center", $client->getRegion("test", "test.alibaba.api.com", null));
        $this->assertEquals("center", $client->getRegion("test", "test.aliyuncs.com", null));
        $this->assertEquals("center", $client->getRegion("test", "test-dualstack.aliyuncs.com", null));
        $this->assertEquals("center", $client->getRegion("test", "test-inner.aliyuncs.com", null));
        $this->assertEquals("center", $client->getRegion("test", "test-vpc.aliyuncs.com", null));
        $this->assertEquals("center", $client->getRegion("test", "test-share.aliyuncs.com", null));
        $this->assertEquals("center", $client->getRegion("test", "test-cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou", $client->getRegion("test", "test.cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou", $client->getRegion("test", "test-inner.cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou", $client->getRegion("test", "test-vpc.cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou", $client->getRegion("test", "test-share.cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou", $client->getRegion("test", "test-dualstack.cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou", $client->getRegion("test", "test-proxy.cn-hangzhou.aliyuncs.com", null));
        $this->assertEquals("cn-hangzhou-acdr-ut-1", $client->getRegion("test", "test-inner.cn-hangzhou-acdr-ut-1.aliyuncs.com", null));
        $this->assertEquals("cn-edge-1", $client->getRegion("test", "test-inner.cn-edge-1.aliyuncs.com", null));
    }

    public function testGetSignedHeaders()
    {
        $client = new Client();

        // 测试空headers
        $emptyHeaders = [];
        $result = $client->getSignedHeaders($emptyHeaders);
        $this->assertEmpty($result);

        // 测试只包含需要签名的headers
        $headers = [
            "host" => "example.com",
            "content-type" => "application/json",
            "x-acs-action" => "TestAction"
        ];
        $result = $client->getSignedHeaders($headers);
        $this->assertCount(3, $result);
        $this->assertEquals("content-type", $result[0]);
        $this->assertEquals("host", $result[1]);
        $this->assertEquals("x-acs-action", $result[2]);

        // 测试包含不需要签名的headers
        $headersWithIgnore = [
            "host" => "example.com",
            "content-type" => "application/json",
            "x-acs-action" => "TestAction",
            "authorization" => "Bearer token",
            "user-agent" => "PHP-client"
        ];
        $result = $client->getSignedHeaders($headersWithIgnore);
        $this->assertCount(3, $result);
        $this->assertEquals("content-type", $result[0]);
        $this->assertEquals("host", $result[1]);
        $this->assertEquals("x-acs-action", $result[2]);

        // 测试空值header
        $headersWithEmpty = [
            "host" => "example.com",
            "content-type" => null,
            "x-acs-action" => "TestAction"
        ];
        $result = $client->getSignedHeaders($headersWithEmpty);
        $this->assertCount(2, $result);
        $this->assertEquals("host", $result[0]);
        $this->assertEquals("x-acs-action", $result[1]);

        // 测试大小写不敏感
        $headersWithCase = [
            "HOST" => "example.com",
            "Content-Type" => "application/json",
            "X-Acs-Action" => "TestAction"
        ];
        $result = $client->getSignedHeaders($headersWithCase);
        $this->assertCount(3, $result);
        $this->assertEquals("content-type", $result[0]);
        $this->assertEquals("host", $result[1]);
        $this->assertEquals("x-acs-action", $result[2]);
    }

}

File: alibabacloud-gateway-pop/csharp/tests/UnitTest.cs
using System;
using System.Collections.Generic;
using AlibabaCloud.GatewayPop;

using Xunit;

namespace tests
{
    public class UnitTest
    {
        [Fact]
        public void Test_GetRegion()
        {
            Client client = new Client();
            Assert.Equal("center", client.GetRegion(null, null, null));
            Assert.Equal("cn-hangzhou", client.GetRegion(null, null, "cn-hangzhou"));
            Assert.Equal("center", client.GetRegion("", "", null));
            Assert.Equal("center", client.GetRegion("test", "test", null));
            Assert.Equal("center", client.GetRegion("test", "test.alibaba.api.com", null));
            Assert.Equal("center", client.GetRegion("test", "test.aliyuncs.com", null));
            Assert.Equal("center", client.GetRegion("test", "test-dualstack.aliyuncs.com", null));
            Assert.Equal("center", client.GetRegion("test", "test-inner.aliyuncs.com", null));
            Assert.Equal("center", client.GetRegion("test", "test-vpc.aliyuncs.com", null));
            Assert.Equal("center", client.GetRegion("test", "test-share.aliyuncs.com", null));
            Assert.Equal("center", client.GetRegion("test", "test-cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou", client.GetRegion("test", "test.cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou", client.GetRegion("test", "test-inner.cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou", client.GetRegion("test", "test-vpc.cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou", client.GetRegion("test", "test-share.cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou", client.GetRegion("test", "test-dualstack.cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou", client.GetRegion("test", "test-proxy.cn-hangzhou.aliyuncs.com", null));
            Assert.Equal("cn-hangzhou-acdr-ut-1", client.GetRegion("test", "test-inner.cn-hangzhou-acdr-ut-1.aliyuncs.com", null));
            Assert.Equal("cn-edge-1", client.GetRegion("test", "test-inner.cn-edge-1.aliyuncs.com", null));
        }

        [Fact]
        public void Test_GetSignedHeaders()
        {
            Client client = new Client();

            // 测试空headers
            var emptyHeaders = new Dictionary<string, string>();
            var result = client.GetSignedHeaders(emptyHeaders);
            Assert.Equal(1, result.Count);

            // 测试只包含需要签名的headers
            var headers = new Dictionary<string, string>
            {
                { "host", "example.com" },
                { "Host", "example.com" },
                { "content-type", "application/json" },
                { "x-acs-action", "TestAction" }
            };
            result = client.GetSignedHeaders(headers);
            Assert.Equal(3, result.Count);
            Assert.Equal("content-type", result[0]);
            Assert.Equal("host", result[1]);
            Assert.Equal("x-acs-action", result[2]);

            // 测试包含不需要签名的headers
            var headersWithIgnore = new Dictionary<string, string>
            {
                { "host", "example.com" },
                { "content-type", "application/json" },
                { "x-acs-action", "TestAction" },
                { "authorization", "Bearer token" },
                { "user-agent", "C#-client" }
            };
            result = client.GetSignedHeaders(headersWithIgnore);
            Assert.Equal(3, result.Count);
            Assert.Equal("content-type", result[0]);
            Assert.Equal("host", result[1]);
            Assert.Equal("x-acs-action", result[2]);

            // 测试空值header
            var headersWithEmpty = new Dictionary<string, string>
            {
                { "host", "example.com" },
                { "content-type", null },
                { "x-acs-action", "TestAction" }
            };
            result = client.GetSignedHeaders(headersWithEmpty);
            Assert.Equal(2, result.Count);
            Assert.Equal("host", result[0]);
            Assert.Equal("x-acs-action", result[1]);

            // 测试大小写不敏感
            var headersWithCase = new Dictionary<string, string>
            {
                { "HOST", "example.com" },
                { "Content-Type", "application/json" },
                { "X-Acs-Action", "TestAction" }
            };
            result = client.GetSignedHeaders(headersWithCase);
            Assert.Equal(3, result.Count);
            Assert.Equal("content-type", result[0]);
            Assert.Equal("host", result[1]);
            Assert.Equal("x-acs-action", result[2]);
        }
    }
}